tensorflow/contrib/eager/python/examples/nmt_with_attention/NMT_with_Attention.ipynb

```diff
@@ -0,0 +1,992 @@
+    },
+    {
+      "metadata": {
+        "id": "nZ2rI24i3jFg",
+        "colab_type": "code",
+        "colab": {
+          "autoexec": {
+            "startup": false,
+            "wait_interval": 0
+          }
+        }
+      },
+      "cell_type": "code",
+      "source": [
+        "class Encoder(tf.keras.Model):\n",
+        "    def __init__(self, vocab_size, embedding_dim, enc_units, batch_sz):\n",
+        "        super(Encoder, self).__init__()\n",
+        "        self.batch_sz = batch_sz\n",
+        "        self.enc_units = enc_units\n",
+        "        self.embedding = tf.keras.layers.Embedding(vocab_size, embedding_dim)\n",
+        "        \n",
+        "        # If you have a GPU, we recommend using CuDNNGRU(provides a 3x speedup than GRU)\n",
+        "        # the code automatically does that.\n",
+        "        if tf.test.is_gpu_available():\n",
+        "          self.gru = tf.keras.layers.CuDNNGRU(self.enc_units, \n",
+        "                                              return_sequences=True, \n",
+        "                                              return_state=True, \n",
+        "                                              recurrent_initializer='glorot_uniform')\n",
+        "        else:\n",
+        "          self.gru = tf.keras.layers.GRU(self.enc_units, \n",
+        "                                         return_sequences=True, \n",
+        "                                         return_state=True, \n",
+        "                                         recurrent_activation='sigmoid', \n",
+        "                                         recurrent_initializer='glorot_uniform')\n",
+        "\n",
+        "    def call(self, x, hidden):\n",
+        "        x = self.embedding(x)\n",
+        "        output, state = self.gru(x, initial_state = hidden)        \n",
+        "        return output, state\n",
+        "    \n",
+        "    def initialize_hidden_state(self):\n",
+        "        return tf.zeros((self.batch_sz, self.enc_units))"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "yJ_B3mhW3jFk",
+        "colab_type": "code",
+        "colab": {
+          "autoexec": {
+            "startup": false,
+            "wait_interval": 0
+          }
+        }
+      },
+      "cell_type": "code",
+      "source": [
+        "class Decoder(tf.keras.Model):\n",
+        "    def __init__(self, vocab_size, embedding_dim, dec_units, batch_sz):\n",
+        "        super(Decoder, self).__init__()\n",
+        "        self.batch_sz = batch_sz\n",
+        "        self.dec_units = dec_units\n",
+        "        self.embedding = tf.keras.layers.Embedding(vocab_size, embedding_dim)\n",
+        "        \n",
+        "        # If you have a GPU, we recommend using CuDNNGRU(provides a 3x speedup than GRU)\n",
+        "        # the code automatically does that.\n",
+        "        if tf.test.is_gpu_available():\n",
```

Comment:
Refactor this into a GRU function so you can remove the duplicates?

Reply:
Done